Fix #8294 push added cookie

gregw opened the pull request:
Reparse cookie added with addCookie via the `java.net.HttpCookie`  for #8294

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Fix #8294 push added cookie

Reparse cookie added with addCookie

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime commented on jetty-server/src/test/java/org/eclipse/jetty/server/RequestTest.java:
```diff
@@ -1965,6 +1966,7 @@ public void testPushBuilder()
         request.getResponse().getHttpFields().add(new HttpCookie.SetCookieHttpField(new HttpCookie("good", "thumbsup", 100), CookieCompliance.RFC6265));
         request.getResponse().getHttpFields().add(new HttpCookie.SetCookieHttpField(new HttpCookie("bonza", "bewdy", 1), CookieCompliance.RFC6265));
         request.getResponse().getHttpFields().add(new HttpCookie.SetCookieHttpField(new HttpCookie("bad", "thumbsdown", 0), CookieCompliance.RFC6265));
+        request.getResponse().getHttpFields().add(new HttpField(HttpHeader.SET_COOKIE, new HttpCookie("ugly", "thumbsdown", 100).getSetCookie(CookieCompliance.RFC6265)));
```
We should get the example cookie that the OP is setting and use it here.
There might be an aspect of RFC6265 we've missed.

joakime replied:
This usage, of having a `HttpField` with an `HttpCookie` object as a value isn't parsing like you think.
Add a Same-Site to that cookie, and see if the resulting cookie from the Push has that Same-Site attribute.

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -352,7 +352,10 @@ public PushBuilder newPushBuilder()
-                HttpCookie cookie = ((SetCookieHttpField)field).getHttpCookie();
+                HttpCookie cookie = (field instanceof SetCookieHttpField)
+                    ? ((SetCookieHttpField)field).getHttpCookie()
+                    : new HttpCookie(field.getValue());
```
This reparse should follow the `HttpConfiguration` setting for `response._responseCookieCompliance` too.

gregw replied:
Hmm maybe.  But that is about the format of the set-cookies that jetty will generate.  This is a set-cookie that the application has generated, probably to get around some compliance issue.
Also, we don't actually have a compliance specific Set-Cookie parser, as we currently only use that part for generation.  So that would be a bit of work.    

So I think this is a step in the right direction, if not perfect.

joakime replied:
The normal cookie parsing step is done via the `CookieCutter` class, which has compliance modes, not the direct `org.eclipse.jetty.http.HttpCookie(String)` constructor, which uses the `java.net.HttpCookie` class to parse, which has limitations (namely it only supports RFC2109 and RFC2965).  

For example, if the OPs `Set-Cookie` has a Same-Site attribute the `org.eclipse.jetty.http.HttpCookie(String)` usage will not parse it.

gregw replied:
CookieCutter only parses Cookie values, not Set-Cookie values... and then it is mostly about ancient long deprecated RFCs.

I've extended the test to make sure we don't throw anything if the same-site attribute is set.  We don't.  It is ignored, but then all we need is max-age to be parsed correctly, which it is.
So I still think we are good.   We don't need to parse same-site, as by definition the cookie on a push request is going to the same site.

joakime replied:
ok, works for me.

joakime reviewed (changes requested):
Reparse should follow the HttpConfiguration configuration for response cookie compliance.

gregw committed:
Fix #8294 push added cookie

Added extra test to ensure maxAge is being parsed

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Fix #8294 push added cookie

Added extra test to ensure maxAge is being parsed with other cookie attributes

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@joakime  can you merge for me and then merge through 11 & 12.
Note where it ends up in 12... Ee9 or 10? Hopefully 9 and then copy to 10

joakime commented:
Merged to `jetty-12.0.x` at ee9 level.
But the ee10 `ServletContextRequest.ServletApiRequest` does not have `newPushBuilder()` implemented yet.
